internal/types/action/registrar.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package action
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/boundary/internal/types/resource"
+)
+
+type resourceActionSets struct {
+	individual ActionSet
+	collection ActionSet
+	valid      ActionSet
+}
+
+type byResource struct {
+	mu sync.RWMutex
+	m  map[resource.Type]*resourceActionSets
+}
+
+func (b *byResource) add(r resource.Type, individual ActionSet, collection ActionSet) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.m == nil {
+		b.m = make(map[resource.Type]*resourceActionSets)
+	}
+
+	_, ok := b.m[r]
+	if ok {
+		return fmt.Errorf("%s already registered", r)
+	}
+	b.m[r] = &resourceActionSets{
+		individual: individual,
+		collection: collection,
+		valid:      ActionSetUnion(individual, collection),
+	}
+
+	return nil
+}
+
+func (b *byResource) get(r resource.Type) (*resourceActionSets, error) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	a, ok := b.m[r]
+	if !ok {
+		return nil, fmt.Errorf("resource not found: %s", r)
+	}
+	return a, nil
+}
+
+var byResourceRegistrar = &byResource{}
+
+// RegisterResource registers ActionSets for r. If RegisterResource is called
+// twice with the same resource, it panics.
+func RegisterResource(r resource.Type, individual ActionSet, collection ActionSet) {
```

Comment:
do we want unit tests for these two exported functions.

Reply:
Added a test in [670d49b](https://github.com/hashicorp/boundary/pull/3825/commits/670d49b35466294b14617cd446a6ec05d44ed3e2) and then expanded on the test some in [cda1217](https://github.com/hashicorp/boundary/pull/3825/commits/cda1217130627c6280d87fdd810cf594f7804401). I looked at adding some additional test cases, like attempting to register and invalid `resource.Type`, but it seems we don't have an existing way to checking if a resource.Type is valid currently. I can look into adding some validation logic for it, but I think that should be a separate PR.